### ISSUE TITLE
instance chains and class constraints

### ISSUE THREAD

**Issue by coot** [purescript/purescript#3120]
It would be nice to be able to chain through failing type class constrains:
```
class IsInt a

instance isIntInt :: IsInt Int

class AsInt a where
  asInt :: a -> Int

instance isIsIntAsInt :: IsInt a => AsInt a where
  asInt a = a
else instance anyAsInt :: AsInt a where
  asInt _ = 0
```

Obviously this could be re-implemented without the `IsInt` tc-constraint, but there are examples where  this will not be that easy.

**Comment by LiamGoodacre** [comment]
In the instance chain paper they discuss doing this kind of thing with a separate instance 'guard'.
But it sounds rather complicated.  I originally didn't implement this to keep it simple.

**Comment by coot** [comment]
I have a use case for it. In [purescript-argonaut-aeson-generic](https://github.com/coot/purescript-argonaut-aeson-generic/blob/allNullaryToStringTag/src/Data/Argonaut/Aeson/Encode/Generic.purs#L54)  I tried to implement aesons `allNullaryAsString` option.

**Comment by LiamGoodacre** [comment]
I've been thinking about a cheaper version of instance guards.  In which we can ask if an instance can be selected for a constraint instead of asking if the constraint holds.  I think that should be fairly quick & simple to implement, and give us most of the benefits.

Doesn't necessarily even need to be part of an instance declaration: could just be a separate syntactic form of a constraint, like a constraint `if (Eq t) b =>` with types `(t :: Type)` & `(b :: Boolean)` and `b` being fully determined.  (Chose `if` due to it already being a reserved word and not valid in this context).

Any thoughts?

**Comment by LiamGoodacre** [comment]
The `if` idea cannot contribute to functional dependencies - because sometimes they are recursive determined.  If we do it as an actual instance guard, then they could influence functional dependencies if we lift them to also be instance constraints.  But then we wouldn't be able to say things like `not (C x) || D x` - i.e form some boolean selection expression.

So I think neither 'contribute to functional dependencies' nor 'lift to constraints' should be done.  They could manually be lifted anyway - in which case, the added constraint can then contribute to functional dependencies.

I'm not too fond of the `if` idea now.  It could be implemented with guards anyway.

**Comment by btford** [comment]
I'm also interested in this– in particular for having more control over custom type errors in the presence of constraints.

@LiamGoodacre – what do you think a solution involving guards look like syntactically?

**Comment by hdgarrood** [comment]
@LiamGoodacre which paper did you read which discusses a separate guard? In the paper [Instance Chains: Type Class Programming Without Overlapping Instances](https://web.cecs.pdx.edu/~mpj/pubs/instancechains.pdf) I don't think they introduce anything in addition to plain old instance contexts; instead, they talk about how they implemented a solver which backtracks when it encounters constraints it can't satisfy:

> 3.1.3 Backtracking search
Haskell instance search never backtracks: if no two instance heads unify,
then no predicate could be solved by more than one instance. However,
combined with overlapping instances, this complicates reasoning about
instances. Even if an instance could apply to a predicate, it will not be
checked if a more specific instance exists anywhere in the program, and
failure to prove the preconditions of the most specific instance causes
instance search to fail rather than to attempt to use less specific
instances. ilab instance search backtracks when it can disprove the
pre-condition of an instance (either because of a fails clause or because
of a functional dependency). When backtracking, ilab checks clauses within
an instance chain in order. The order in which ilab checks instance chains
is unimportant because clauses in different chains are not allowed to
overlap.

This, however, is not something either @natefaubion or I are very enthusiastic about (see #3596).

@coot Do you know how Aeson implements `allNullaryAsString`? If I'm reading it correctly, the above quote implies that the feature being requested here isn't actually necessary to do it in Haskell.

**Comment by LiamGoodacre** [comment]
Yeah I probably just misremembered it.  But either way, I'm not very keen on having instance search backtrack on unsatisfiable contexts.

**Comment by kl0tl** [comment]
Backtracking as presented in the paper by Morris crucially depends on _negative predicates_. Failure to prove a constraint isn’t enough to backtrack, a proof that the constraint fails is required. Such proof can be deduced (for example we can disprove `Elem (Array a) Int` if we can prove `Elem (Array a) a`, given `class Elem xs x | xs -> x`) or explicitely introduced by failing instances (`instance C a fails` in the paper).

This was perhaps already obvious but that’s not what I understood from this thread and https://github.com/purescript/purescript/issues/3596 before carefully reading the paper and skimming through Morris PhD dissertation, so I figured it would be good to restate things here.
